edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/NestedCredentialCache.cs

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.Core
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Identity;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Azure.Devices.Edge.Util.Concurrency;
+
+    public class NestedCredentialsCache : ICredentialsCache
+    {
+        // Maps client identity -> credential used for auth
+        readonly IDictionary<string, IClientCredentials> credentials = new Dictionary<string, IClientCredentials>();
+
+        // Maps OnBehalfOf actor device -> children, e.g. edge1 -> [leaf1, edge1/$edgeAgent, ...]
+        readonly IDictionary<string, HashSet<string>> onBehalfOfRelations = new Dictionary<string, HashSet<string>>();
+
+        readonly ICredentialsCache underlyingCache;
+        readonly AsyncLock cacheLock = new AsyncLock();
+
+        public NestedCredentialsCache(ICredentialsCache underlyingCache)
+        {
+            this.underlyingCache = Preconditions.CheckNotNull(underlyingCache, nameof(underlyingCache));
+        }
+
+        public async Task Add(IClientCredentials clientCredentials)
+        {
+            using (await this.cacheLock.LockAsync())
+            {
+                // Update the in-memory relations
+                this.UpdateCachedCredentials(clientCredentials);
+
+                // Update the DB
+                await this.underlyingCache.Add(clientCredentials);
+            }
+        }
+
+        public async Task<Option<IClientCredentials>> Get(IIdentity identity)
+        {
+            using (await this.cacheLock.LockAsync())
+            {
+                if (!this.credentials.TryGetValue(identity.Id, out IClientCredentials clientCredentials))
+                {
+                    // Update the in-memory relations as we read credentials out of the DB
+                    Option<IClientCredentials> underlyingStoreCredentials = await this.underlyingCache.Get(identity);
+                    underlyingStoreCredentials.ForEach(creds => this.UpdateCachedCredentials(creds));
+                    return underlyingStoreCredentials;
+                }
+
+                return Option.Some(clientCredentials);
+            }
+        }
+
+        void UpdateCachedCredentials(IClientCredentials clientCredentials)
+        {
+            string targetId = AuthChainHelpers.GetAuthTarget(clientCredentials.AuthChain).GetOrElse(clientCredentials.Identity.Id);
+
+            if (clientCredentials.AuthChain.HasValue)
```

Comment:
better way to do this is clientCredentials.AuthChain.Map().OrElse(), or clientCredentials.AuthChain.Match()

Reply:
both approach requires a return value in the lambdas, which doesn't really apply to either if/else block.  If I make them return some arbitrary type then it prob looks more awkward =(